### PR TITLE
Refactor Hollows solvers, add tests, add Vec3Comparable, fix bugs

### PR DESCRIPTION
**Pull request by CraftyOldMiner**
This PR includes the following changes:

**Metal Detector:**
- Refactored to be unit testable in support of making future fixes for sporadic failures with minimal regression risk.
- Added extensive unit tests that allow the core logic to be debugged without launching Minecraft.
- Added debug option to turn off usage of center coordinates

**Wishing Compass**
- Rewrote the Wishing Compass Solver to be unit testable.
- Solver now determines target type based on found crystals, inventory, and coordinates.
- Jungle Temple target coordinates adjust to point at temple door.
- Detect and handle the second compass having different possible targets than the first.
- Fixed bugs that led to incorrect solutions and the solver getting stuck in a broken state.
- Block using the wishing compass in the Nucelus.
- Allow the second wishing compass to be used as soon as the first compass' initial sequence has finished.
- Added extensive unit tests that allow the core logic to be debugged without launching Minecraft.
- Added options to auto-create Skytils waypoints with either Skytils built-in names or separate names (to avoid Skytils overwriting them).
- Made non-automatic option to add a Skytils waypoint more obvious.

**Other Items**
- Added a Vec3Comparable util class with test cases
- Moved logic for checking that a SkyBlock location is available into SBInfo for shared use

### CODE REVIEW

File: src/main/java/io/github/moulberry/notenoughupdates/core/util/Vec3Comparable.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		return new Vec3Comparable(super.rotatePitch(pitch));
+	}
+
+	@Override
+	public Vec3Comparable rotateYaw(float yaw) {
+		return new Vec3Comparable(super.rotateYaw(yaw));
+	}
+
+	@Override
+	public boolean equals(Object other)	{
```
Comment (romangraef) [inline src/main/java/io/github/moulberry/notenoughupdates/core/util/Vec3Comparable.java:86]: This method technically breaks the symmetry contract of equals: 
```java
var a = new Vec3(0,0,0);
var b = new Vec3Comparable(0,0,0);
a.equals(b) // false
b.equals(a) // true```

Comment (CraftyOldMiner) [reply src/main/java/io/github/moulberry/notenoughupdates/core/util/Vec3Comparable.java:86]: Good catch - I'll fix it.

**Comment by CraftyOldMiner** [comment]
FYI - I am finishing up testing some changes that I will be updating this PR with tomorrow. The changes identify the kind of waypoint that the compass is pointing to and allows for auto-creation.

**Review by NopoTheGamer — APPROVED** [review]
math :+1:

**Review by NopoTheGamer — APPROVED** [review]
:+1: hypxiel

**Review by Lulonaut — COMMENTED** [review]
Is there a reason to have this file: src/test/resources/log4j2-test.xml ?
I don't see it being used anywhere and its note mentioned when running the tests.

**Comment by CraftyOldMiner** [comment]
> Is there a reason to have this file: src/test/resources/log4j2-test.xml ? I don't see it being used anywhere and its not mentioned when running the tests.

See https://howtodoinjava.com/log4j2/configure-log4j2-for-junit/ & https://logging.apache.org/log4j/2.x/manual/configuration.html#AutomaticConfiguration. Without this config file, running unit tests will result in the CrystalHollowsMetalDetector static class writing an empty file at <git root>\logs\latest.log.

**Review by Lulonaut — APPROVED** [review]

**Review by romangraef — APPROVED** [review]
lgtm

**Review by NopoTheGamer — APPROVED** [review]
CraftyOldMiner dismissed NopoTheGamer's stale review :-1:

**Pull request merged**